### DISCUSSION TITLE
lrzip and squashfs, mount, block based online decompression

### DISCUSSION THREAD

**Discussion by lukypko** [pete4abw/lrzip-next#82] [Ideas]
Hello,
Would it be possible to implement an **squashfs like** block based, random, online decompression. I mean you can define a minimal block size which can be decompressed as you want to read a single byte from it.
- then it would be possible to mount an *lrzip* file and decompress online just random parts of files which are requested
- if needed it can use some kind of disk/memory cache to store some helping data
- I really like **rzip** long range lookup, because it significantly reduces output file size as you compress a similar data
- compression level is not so important, it can be **zstd** too, as it is really fast

Is it technically possible to use an advantage of *rzip* to random decompression of any part of output file to be able to mount it?
Is it possible to change a format to allow that?
Thank you :-)

**Comment by pete4abw** [comment]
Thanks for the suggestion. There a number of issues here.
1. The library for lrzip-next has been removed.
2. lrzip/lrzip-next is not a journaling archiver. It works on one file at a time.
3. lrzip/lrzip-next is used for larger files, whereas files in a ramdisk may be small or large.
4. lrzip/lrzip-next is not really fast when compared to squashfs.
5. lrzip-next does not support zstd and there are no plans to.

I know where you want to go with this, open an archive in a file manager, `mc`, etc., and view/extract the contents interactively.

If you wish to contribute code to support this issue, please feel free.

Thank you again!

**Comment by lukypko** [comment]
Hello,
after investigation of format and how **rzip** works in general, **I have added a support for multiple files/folders**. Technically, all what I did is mapping a multiple input/output files into a single concatenated stream and allow to seek in it.
- it is possible to add any count of files/folders into an lrzip archive, all files are processed in a single chunk, so if there are same parts in a first and in the last file, they should be caught by a rzip algorithm and stored just once
- it is possible to decompress a whole archive only, so it is not possible to decompress just a one file
- works only for 64 bit operating system, because I'm using *mmap*, It is possible add a support for 32-bit too
- removed support for compressing of *stdin*, because it was just a "fake" (so it does not compressed a whole input stream, but just compressed a small chunk loaded into a memory)
- performance of compression and decompression is similar like just compression of a single file
- works only on Linux :-)


Now, I have a plan to implement:
- cleanup and share a code (I have formatted all files in IDE, so I have to exclude just my changes)
- decompress just a smaller part of the file, decompress just files which match the filter (regular expression)
- mounting of lrzip/rzip archive, theoretically all unique parts of file are just stored once, so maybe it will not be catastrophically slow to decompress a random range of a file, at first I have to implement it and then "we will see"
- adding files into archive incrementally, but I do not understand algorithm so deeply, maybe it is possible to regenerate (or store) those hashes and add new files without recompression of already compressed files
- rename already stored files in archive without a recompression
- "in-place" compression, so imagine you want to compress a 100GB of data but have only 1 GB of free disk space; already processed input data will be "holed", make parts of input data sparse, so it would not consume a disk space

Generic thoughts:
- I have minimal skills in C language, it is pain to allocate and then do not forget to deallocate, pain to work without exceptions,
  maybe C++ or rust can be better for that

- how many people know about rzip/lrzip/lrzip-next?
- how many people are practically using rzip/lrzip/lrzip-next to compress their files? I still consider this project as a very theoretical.

**Reply by pete4abw** [reply]
@lukypko . Thank you for all the effort you put into this discussion! I'll study this in more detail, but want to give you my off the top of my head feedback.
* The inability to extract one or more files is a handicap. There has to be a method to list all files in the archive and then pick and choose.
* Removing stdin/stdout will be a real problem for users.
* Using
`tar --use-compress-program=lrzip-next -cf file.tar.lrz file1 file2 file3...` and
`tar --use-compress-program=lrzip-next -x|t -f file.lrz`
really can accomplish journaling and selective file extract.
* renaming of stored files is very risky because of integrity hashing which might fail.
* I don't understand the "in-place" compression feature. Most systems today have lots of memory and lots of storage.
* i don't know how many people use `lrzip/lrzip-next`. It's a unique program with a unique following. Probably not many.
* While I do not agree that the project is "theoretical" what @ckolivas created was a very unique way to extend the benefits of `rzip` into standard file compression. Because neither `lrzip/lrzip-next` is considered in-production (version < 1.0), it should technically be considered **beta**. However, this program has been in development and used since at least 2007. I have been involved since then too. It's files are hash-checked and verified (in the case of `lrzip-next`) so the integrity of the data is assured. There are no outstanding CVEs against it. So it is much more than theoretical!

Thank you again for your feedback. I'll keep checking your suggestions, but my initial reaction is NOT to add bloat to `lrzip-next`. Maybe it IS easy to do, but I don't see it that way and there is already a way to do this with `tar` or even `dd`.

**Comment by lukypko** [comment]
Hello,
finally I made a fork of this project and have commit my changes, they are in https://github.com/lukypko/lrzip-next, they are in a branch `files1`.

- it has support for compression and decompression of multiple files
- theoretically it is possible to make support for *stdin/stdout*, but I have to better understand to rzip algorithm
- it is a **live version** which I plan to modify (if I have a time for that)

lrzip archive files are stored in a external file :-), it is just temporary unless I find a better format

Please feel free to test it, use **firejail** or something...

PS: it would be great if somebody could write a detail description of *rzip algorithm.*

**Comment by pete4abw** [comment]
Thank you for your efforts.
A lot of things here. 
* `lrzip` and `lrzip-next` were never intended to be journalling archivers.
* `lrzip-next` can compress multiple files at a time, but no guarantees you won't run out of memory! It just won't concatenate the output. Try
`lrzip-next *.c`  and `lrzip -d *.c.lrz` for example.
 -rw-r--r-- 1 peter users  14808 Feb  3 07:33 /tmp/lrzip.c.lrz
 -rw-r--r-- 1 peter users   9490 Feb  3 07:33 /tmp/main.c.lrz
 -rw-r--r-- 1 peter users   5108 Jun 21  2022 /tmp/runzip.c.lrz
 -rw-r--r-- 1 peter users  11092 Nov 18 11:19 /tmp/rzip.c.lrz
 -rw-r--r-- 1 peter users  16951 Feb  3 07:36 /tmp/stream.c.lrz
 -rw-r--r-- 1 peter users   5806 Feb  3 07:33 /tmp/util.c.lrz
* stdin/stdout are already supported for single files.
* `tar` is best used if you want multiple files or directories. It works great and no reason to change.
* I encourage you to keep working on your fork and keep us informed of your progress.

RZIP info can be viewed here: [Andrew Tridgell Thesis: Efficient Algorithms for Sorting and
Synchronization](https://www.samba.org/tridge/phd_thesis.pdf). See Chapter 5 beginning on page 84.

You may also reference [rzip page](https://rzip.samba.org/) and [rzip64](http://rzip64.ghsi.eu/).
